active/0004-extensibility-support-for-deployment-stacks.md

```diff
@@ -0,0 +1,665 @@
+    }
+  }
+  params: {
+    // ...
+  }
+}
+```
+
+extResources.bicep - Child deployment with extension resources
+```bicep
+extension kubernetes as k8s // optional alias
+
+extension microsoftGraph // No config is needed for this one. This extension uses OAuth OBO.
+
+// ... extension resources
+```
+
+The keys of the "extensions" object are the extension aliases of the module deployment. If an alias is not provided,
+it will be the extension name. The values of the object are extension configurations. Optionality of configuration is
+determined by the extension. If this is compared to a typical resource group deployment, it is aligned because the
+resource group being deployed to is not defined within the deployment itself, but as a parameter to the deployment,
+either through the CLI parameters or as a scope for nested deployments. In this example, the deployment scope(s) are 
+being defined as extension configurations. Expressions in this object can be constrained to expressions that can be 
+compiled to simple directives that both the deployments engine and the stacks service can execute.
+
+Allowed expressions:
+- ARM resource access that can be compiled to a resource ID (`aksCluster` in the example).
+- ARM resource API calls that can be compiled to a URI, HTTP method, and any inputs (`listClusterAdminCredential` in the
+  example).
+- JSON value accessors (`kubeconfig[0].value` in the example).
+- Usage of compile-time constants (the `0` in `kubeconfig[0]` in the example).
+- Usage of non-secure deployment parameters (`namespace` value in the example).
+
+Disallowed expressions:
+- Usage of secure deployment parameters.
+- For configuration properties that are marked as secret/secure (`kubeConfig` property in the example):
+    - Any expression that is not direct access on an ARM resource reference or complicates the analysis.
+
+The reasoning behind the disallowed expressions is because the "extensions" property will be evaluated by the deployment
+engine and the evaluation results will be returned in the deployment GET response. The stacks service will persist this
+data long-term and use it for the next stack update or deletion that requires extension resource deletions. These
+constraints will make it required that secret properties ("kubeConfig" in the Kubernetes example) are provided a value
+in an expected fashion like ARM resource APIs. It also prohibits usage of Bicep functions such as `loadFileContent` to
+inline load secret content.
+
+Examples of expression allowance for the `kubeConfig` case:
+```bicep
+@secure()
+param secureParam string
+
+param nonSecureParam string
+
+param boolParam bool
+
+resource aksCluster '...' = {
+  // ...
+}
+
+module extResources 'extResources.bicep' = {
+  name: 'extResourcesDeploymentName'
+  extensions: {
+    k8s: {
+     ✅ kubeConfig: aksCluster.listClusterAdminCredential().kubeconfig[0].value
+     ❌ kubeConfig: loadFileContent('kubeconfig.txt')
+     ❌ kubeConfig: secureParam
```

Comment:
I believe the two use cases mentioned can still be valuable for non-stack deployments, and customers might still want to use them. Rather than disallowing them, could we introduce a linter rule to indicate that they won't function with stacks?

Reply:
Added linter rule design.